sklearn/base.py

```diff
@@ -556,6 +582,15 @@ def fit_transform(self, X, y=None, **fit_params):
             return self.fit(X, y, **fit_params).transform(X)
 
 
+class NonRectangularInputMixin:
+    """Mixin class for all estimators with non-rectangular input.
+
+    For now only vectorizers are relevant for this mixin.
```

Comment:
Meta-estimators may also sometimes receive non-rectangular input. I think you need to think through the meta-estimator case a bit more.

LabelEncoder and LabelBinarizer might also qualify... but they're weird for several reasons.

Reply:
I think the meta-estimators require extra code to delegate to the base estimator. meta-estimators ideally shouldn't do any input validation so they won't try to set it themselves.